[MIG][15.0] hr: migration to 15.0

daonamutc1 opened the pull request:
[Task](https://viindoo.com/web#id=22879&action=409&active_id=177&model=project.task&view_type=form&cids=1&menu_id=289)
- Convert text to html
- Cretate records for hr.work.location
- Map departure_reason_id
Video:
  - Before: https://drive.google.com/file/d/1qM32JRj5X0o-XozbtFN-XIKFlJH4Fzni/view?usp=sharing
  - After: https://drive.google.com/file/d/1msIKDgeY_0jdAJxYYcxRPkhUtZkMQZnb/view?usp=sharing

royle-vietnam commented on openupgrade_scripts/scripts/hr/15.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,68 @@
+def migrate(env, version):
+    compute_work_location_id(env)
+    compute_departure_reason(env)
+    convert_text_to_html(env)
```
@daonamutc1 anh chuyển hàm `convert_text_to_html` sang `pre-migration` nhé

royle-vietnam commented on openupgrade_scripts/scripts/hr/15.0.1.1/upgrade_analysis_work.txt:
```diff
@@ -0,0 +1,73 @@
+
+hr           / hr.employee              / departure_reason (selection)  : DEL selection_keys: ['fired', 'resigned', 'retired']
+hr           / hr.employee              / departure_reason_id (many2one): NEW relation: hr.departure.reason
+# DONE post-migration: map hr.employaee with hr.deprture.reason
```
```suggestion
# DONE post-migration: map departure_reason with hr.deprture.reason on hr.employee model
```

royle-vietnam commented on openupgrade_scripts/scripts/hr/15.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,68 @@
+        ]
+
+
+def compute_departure_reason(env):
```
```suggestion
def _map_hr_employee_departure_reason_to_departure_reason_id(env):
```

royle-vietnam commented on openupgrade_scripts/scripts/hr/15.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,68 @@
+                    WHEN 'resigned' THEN %s
+                    WHEN 'retired' THEN %s
+                END
+            WHERE departure_reason IS NOT NULL
```
@daonamutc1 tránh trường hợp module khác bổ sung thêm tùy chọn khác
```suggestion
            WHERE departure_reason IN ('fired', 'resigned', 'retired')
```

royle-vietnam commented on openupgrade_scripts/scripts/hr/15.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,60 @@
+from openupgradelib import openupgrade
+
+
+def compute_work_location_id(env):
+    env.cr.execute("SELECT * FROM hr_employee WHERE work_location IS NOT NULL")
+    all_employees = env.cr.dictfetchall()
+
+    work_location_dict = {}
+    for employee in all_employees:
+        if employee["work_location"] not in work_location_dict.keys():
+            address_id = employee.get("address_id", False)
+            if not address_id:
+                address_id = (
+                    env["res.partner"]
+                    .create(
+                        {
+                            "name": employee["work_location"],
+                            "company_type": "person",
+                            "type": "private",
+                        }
+                    )
+                    .id
+                )
+            work_location = env["hr.work.location"].create(
+                {
+                    "name": employee["work_location"],
+                    "company_id": employee["company_id"],
+                    "address_id": address_id,
+                }
+            )
+            work_location_dict[employee["work_location"]] = work_location
+        env["hr.employee"].browse(employee["id"]).work_location_id = work_location_dict[
+            employee["work_location"]
+        ]
```
@daonamutc1 anh test giúp em đoạn này nhé
```suggestion
def _map_hr_employee_work_location_to_work_location_id(env):
    env.cr.execute("""
        WITH work_location_tmp AS (
            INSERT INTO hr_work_location (name, company_id, address_id)
            SELECT emp.work_location, emp.company_id, emp.address_id
            FROM hr_employee emp
            WHERE emp.address_id IS NOT NULL AND emp.work_location IS NOT NULL
            RETURNING id, emp.id as emp_id
        )
        UPDATE hr_employee emp
        SET work_location_id = work_location_tmp.id
        FROM work_location_tmp
        WHERE work_location_tmp.emp_id = emp.id
```

daonamutc1 committed:
[MIG] hr: migration to 15.0
- Convert text to html
- Cretate records for hr.work.location
- Map departure_reason_id
